src/protocol/requests/metadata/topic/topics-factory.ts

```diff
@@ -0,0 +1,58 @@
+import { type UUID } from 'crypto';
+import { CompactNullableString, CompactString, String, Uuid } from '../../../primitives';
+import { TopicV10 } from './topic-v10';
+import { TopicV9 } from './topic-v9';
+
+export type Topics = string[] | { id: UUID; name?: string }[];
+
+export class TopicsFactory {
+  constructor(private readonly topics: Topics) {}
+
+  public createStrings(): String[] {
+    if (this.isStringArray(this.topics)) {
+      return this.topics.map((topic) => new String(topic));
+    }
+
+    if (this.topics.every((topic): topic is { name: string; id: UUID } => topic.name !== undefined)) {
+      return this.topics.map((topic) => new String(topic.name));
+    }
+
+    throw new Error('Invalid topics');
+  }
+
+  public createTopicsV9(): TopicV9[] {
+    if (this.isStringArray(this.topics)) {
+      return this.topics.map((topic) => new TopicV9(new CompactString(topic)));
+    }
+
+    if (this.topics.every((topic): topic is { name: string; id: UUID } => topic.name !== undefined)) {
+      return this.topics.map((topic) => new TopicV9(new CompactString(topic.name)));
+    }
+
+    throw new Error('Invalid topics');
+  }
+
+  public createTopicsV10(): TopicV10[] {
+    if (this.isStringArray(this.topics)) {
+      return this.topics.map((topic) => new TopicV10(Uuid.Zero, new CompactNullableString(topic)));
+    }
+
+    if (this.topics.every((topic): topic is { name: string; id: UUID } => topic.name !== undefined)) {
```

Comment:
This check can also benefit from being extracted to dedicated method, like `isStringArray`.